### PR TITLE
nixos/borgbackup: Document wrapper script and show excludes file path

### PR DESCRIPTION
**Pull request by rvl**
###### Motivation for this change

I wrote my own `borg` wrapper scripts before realising that they are provided by the NixOS module.

This adds documentation of the wrapper scripts.

It also adds the path to the excludes file as a variable, because that is useful to know when testing filters.

###### Things done

- Built on platform(s)
   - [x] NixOS
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rvl** [comment]
Hi @dotlambda, @yorickvP, @ngiger, would you please be able to review this?

File: nixos/modules/services/backup/borgbackup.nix [hunk 1]
```diff
@@ -112,7 +116,7 @@ let
   mkBorgWrapper = name: cfg: mkWrapperDrv {
     original = "${pkgs.borgbackup}/bin/borg";
     name = "borg-job-${name}";
-    set = { BORG_REPO = cfg.repo; } // (mkPassEnv cfg) // cfg.environment;
+    set = mkBorgEnvironment cfg;
```
Comment (dotlambda) [inline nixos/modules/services/backup/borgbackup.nix:119]: This doesn't always make sense: What if the user that runs the wrapper doesn't have permission to execute `passCommand`?

**Review by dotlambda — COMMENTED** [review]
Sorry for taking so long to reply.

I must say I'm not a big fan of the wrapper script and wouldn't have merged the PR adding it.

**Comment by rvl** [comment]
Thanks for the feedback @dotlambda.

The wrapper script was a good addition because it provides a convenient way for users to test their backups.
As you know, a backup can't be trusted unless it's regularly tested.
If there is a better way of testing backups made by the borgbackup module, please let us know and we can document it.

File: nixos/modules/services/backup/borgbackup.nix (hunk 1 above)
Comment (rvl) [reply nixos/modules/services/backup/borgbackup.nix:119]: Same question applies to `BORG_REPO`/`BORG_RSH` - what if the user that runs the wrapper can't access the ssh key?
The wrapper script isn't meant to work when run by any user - that's not its purpose.
Users would have to run it as `config.services.borgbackup.jobs.<name>.user`, which is fine.

File: nixos/modules/services/backup/borgbackup.xml [hunk 2]
```diff
@@ -117,18 +117,27 @@ ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAID78zmOyA+5uPG4Ot0hfAy+sLDPU1L4AiIoRYEIVbbQ/
   };
 };</programlisting>
   </para>
-  <para>The following few commands (run as root) let you test your backup.
-      <programlisting>
-> nixos-rebuild switch
+  <section xml:id="opt-services-backup-borgbackup-testing">
+   <title>Testing remote backups</title>
+   <para>Every job has a corresponding wrapper script installed into <xref linkend="opt-environment.systemPackages"/>. Inspect the wrapper variables by reading <filename>/run/current-system/sw/bin/borg-job-&lt;NAME&gt;</filename>. This script can be used to test backups or manage your remote repo.
```
Comment (dotlambda) [inline nixos/modules/services/backup/borgbackup.xml:122]:
```suggestion
   <para>Every job has a corresponding wrapper script installed into <xref linkend="opt-environment.systemPackages"/>.
   It sets the same environment variables as the systemd unit.
   You can inspect them by reading <filename>/run/current-system/sw/bin/borg-job-&lt;name&gt;</filename>.
   This script can be used to test backups, manage the remote repo, and do restores.
   It is only guaranteed to work if run as the same user as the systemd unit.
```

File: nixos/modules/services/backup/borgbackup.xml [hunk 3]
```diff
@@ -117,18 +117,27 @@ ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAID78zmOyA+5uPG4Ot0hfAy+sLDPU1L4AiIoRYEIVbbQ/
...
+   </para>
+   <para>The following few commands (run as root) show how to test your backup.
```
Comment (dotlambda) [inline nixos/modules/services/backup/borgbackup.xml:124]:
```suggestion
   <para>The following few commands show how to test your backup.
```
A `#` should be clear enough.

File: nixos/modules/services/backup/borgbackup.xml [hunk 4]
```diff
@@ -117,18 +117,27 @@ ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAID78zmOyA+5uPG4Ot0hfAy+sLDPU1L4AiIoRYEIVbbQ/
...
+<screen>
+<prompt># </prompt> nixos-rebuild switch
 ...restarting the following units: polkit.service
-> systemctl restart borgbackup-job-backupToLocalServer
-> sleep 10
-> systemctl restart borgbackup-job-backupToLocalServer
-> export BORG_PASSPHRASE=topSecrect
-> borg list --rsh='ssh -i /run/keys/id_ed25519_my_borg_repo' borg@nixos:.
+<prompt># </prompt> systemctl restart borgbackup-job-backupToLocalServer
```
Comment (dotlambda) [inline nixos/modules/services/backup/borgbackup.xml:128]:
```suggestion
<prompt># </prompt> systemctl start borgbackup-job-backupToLocalServer
```
Why restart?

Comment (rvl) [reply nixos/modules/services/backup/borgbackup.xml]: No idea - that's how it was before. `systemctl start` would be fine.

**Comment by rvl** [comment]
Thanks for the review @dotlambda - I have applied your suggestions.

File: nixos/modules/services/backup/borgbackup.nix [hunk 5]
```diff
@@ -91,13 +91,17 @@ let
...
-      environment = {
-        BORG_REPO = cfg.repo;
+      environment = mkBorgEnvironment cfg // {
         inherit (cfg) extraArgs extraInitArgs extraCreateArgs extraPruneArgs;
-      } // (mkPassEnv cfg) // cfg.environment;
+      };
       inherit (cfg) startAt;
     };
 
+  mkBorgEnvironment = cfg: {
```
Comment (dotlambda) [inline nixos/modules/services/backup/borgbackup.nix:100]:
```suggestion
  mkEnvironment = cfg: {
```
But actually, I don't know why I chose to let those functions have an argument `cfg`. That can just be dropped.

**Comment by mkaito** [comment]
@yorickvP didn't we merge the restore unit and env file we generate off our borg stuff?

**Comment by yorickvP** [comment]
Yeah, this PR documents the generated wrappers.

**Comment by mkaito** [comment]
> I must say I'm not a big fan of the wrapper script and wouldn't have merged the PR adding it.

Please revert it if you don't want it here. I stopped caring about contributing to nixpkgs a long time ago.

**Comment by dotlambda** [comment]
> > I must say I'm not a big fan of the wrapper script and wouldn't have merged the PR adding it.
> 
> Please revert it if you don't want it here. I stopped caring about contributing to nixpkgs a long time ago.
> 
> In case you wonder why, well, dude took time out of his day to document a thing and contribute said documentation for everyone else here, and your response is "wouldn't have merged this in the first place". What an attitude to have.

I just mentioned my opinion, asked a valid question at the same time, and gave (hopefully) valuable feedback later. It's normal that people criticise but -- that's how I know Nixpkgs -- a more or less democratic compromise is found eventually. This is how every open source project should work and we would appreciate your input just as much as anyone else's.

**Review by mkaito — APPROVED** [review]
As far as this PR is concerned, I think we can all appreciate things getting better documentation. Looks good to me.

**Comment by dotlambda** [comment]
> And again, feel free to file a PR reverting everything about this script if you don't like it.

I have no intention to do so, since multiple people seem to find this wrapper script useful.



> The only reason it's even here is that we had some misguided and deluded attempt to upstream all our nixpkgs modifications some time ago. Don't worry, it won't happen again.

I'm sure it would make your life easier if you did upstream at least some of them. And it makes the reviewers' lives easier if you do so in little bits. As I bet other people would profit from these changes, I would love to see you change your mind. If you do, make sure to ping me in those PRs and I'll be happy to drop a review!

File: nixos/modules/services/backup/borgbackup.nix (hunk 5 above)
Comment (rvl) [reply nixos/modules/services/backup/borgbackup.nix]: OK fixed.

`cfg` is not actually the service module config, but the per-repo config. So it needs to be a variable one way or another.

**Comment by rvl** [comment]
@mkaito @yorickvP Your efforts are appreciated by me, at least.

**Comment by mkaito** [comment]
Worth making clear that I don't speak for Yorick, given that I've probably upset a few people.

**Comment by rvl** [comment]
Ready to merge - @Ma27 - thanks.

**Comment by dotlambda** [comment]
@rvl WOuld you be okay with removing the `$excludeFile` hack? I think we should first add such an environment variable upstream.

**Comment by rvl** [comment]
No @dotlambda, that's not reasonable - I'm not going to ask upstream to start making environment variables for every CLI option.
We can and should fix it in our NixOS service scripts.

**Comment by dotlambda** [comment]
Okay, fine. But at least squash the last two commits into one.

**Comment by mkaito** [comment]
For what it's worth, I don't think having environment variables for every option is unreasonable. Perhaps upstream has strong opinions on the subject, but it's not a fundamentally wrong course of action, especially for things meant to run automatically on a server. Environment for configuration is one of the tenets of 12-factor apps, after all.

**Comment by rvl** [comment]
OK thanks @dotlambda - rebased and squashed.

It's not unreasonable to have environment variables for every CLI option, and sometimes it's quite helpful. But most of the software we package does not work like this. Borgbackup has a handful of settings which can be controlled by env var - generally things that apply to every `borg` subcommand. Our purpose is to package the software and make it easy to configure and use, not to fix every upstream package.

**Comment by rvl** [comment]
Hello @dotlambda or @SuperSandro2000, I just rebased this on the latest `nixos-unstable`. Could we please merge it? :pray:

Also in related borgbackup news, I have recently been using the `borgbackup create --patterns-from=patterns.lst` option, which seems like a much more sane way of specifying files to backup. I could add an option to the nixos module for this - if someone is willing to merge it.

**Comment by mkaito** [comment]
This PR was recently brought to my attention again, and I would like to help to get it merged. If @dotlambda doesn't have the time, I can poke a few others and see if they would like to help.

> I have recently been using the `borgbackup create --patterns-from=patterns.lst` option

That's a good idea, but perhaps you should put that in a separate PR (please ping me).

**Comment by SuperSandro2000** [comment]
Since I have zero knowledge about borgbackup someone else needs to verify the changes.

**Pull request closed without merging**